### ISSUE TITLE
Support NumPy 2.0

### ISSUE THREAD

**Issue by kmaehashi** [cupy/cupy#8306]
Make CuPy APIs compatible with NumPy 2.0 🚀 Ref: #8269

Major tasks include:

* New Type Promotion Rules for Python scalars ([NEP 50](https://numpy.org/neps/nep-0050-scalar-promotion.html))
* Python API Cleanup for NumPy 2.0 ([NEP 52](https://numpy.org/neps/nep-0052-python-api-cleanup.html))
* Adopting Array API Compatibility in the main NumPy namespace ([NEP 56](https://numpy.org/neps/nep-0056-array-api-main-namespace.html))

**Comment by rgommers** [comment]
Just to make sure: the plan is for 14.0.0 to completely match NumPy 2.0? Which probably means only testing against NumPy 2.0 in all CI, since there is no way to match 2.0 and 1.26 at the same time?

**Comment by kmaehashi** [comment]
Yes, our current plan is to match CuPy v14 API with NumPy 2.0, and only test against NumPy 2.0+.

**Comment by ev-br** [comment]
I'm going to work towards this, starting with the python api, NEP 52.

**Comment by kmaehashi** [comment]
Regarding NumPy 2.0 support, CuPy currently:

1. requires [`numpy>=1.22,<1.29`](https://github.com/cupy/cupy/blob/6eb9fddaa842c67fe46df88b5444f59147a33102/setup.py#L26), and
2. imports APIs to be removed in NumPy 2.0.

This means users cannot install/import CuPy in NumPy v2 environment.

I'm thinking it's better to ship CuPy v13.2.0 release that includes:

1. updated requirement `numpy>=1.22,<2.3`
2. make imports NumPy 2.0-safe, so that CuPy v13 can be imported with both NumPy v1/v2 environment

before the NumPy 2 ship date.

**Comment by rgommers** [comment]
That would be great. I think not matching exactly all the promotion semantics and missing a few new functions is fine, as long as it imports correctly indeed.

**Comment by kmaehashi** [comment]
We have received several bug reports regarding NumPy v2 support in CuPy v13. They will be tracked under [np2-v13](https://github.com/cupy/cupy/labels/np2-v13 ) label.

**Comment by agm-eratosth** [comment]
Per https://github.com/rapidsai/build-planning/issues/38 this should be supported in 24.10

**Comment by jakirkham** [comment]
> Per https://github.com/rapidsai/build-planning/issues/38 this should be supported in 24.10

So that issue only concerns RAPIDS and its ability to support NumPy 2. Not other libraries (like CuPy). Though RAPIDS does track dependency support

For that reason, 24.10, which is a RAPIDS version, does not mean anything in the context of CuPy, which uses an entirely different versioning scheme

With that out of the way, let's discuss what this issue represents

<hr>

There are two different questions tracked by this CuPy issue:

1. Can CuPy run with NumPy 2.0?
2. Does CuPy's API mirror that of NumPy 2.0?

The answer to 1 is yes. Though a few bug fixes are found from time to time, which are tracked here

The answer to 2 is not until CuPy 14, which is why this issue is still open

<hr>

Hope that helps clarify things 🙂

**Comment by kmaehashi** [comment]
@asi1024 I think the only remaining tasks is `Removes skips with XXX comments`? In that case I think it's better to close this one and open another one to track it, to avoid any confusion.

**Comment by kmaehashi** [comment]
> [@asi1024](https://github.com/asi1024) I think the only remaining tasks is `Removes skips with XXX comments`? In that case I think it's better to close this one and open another one to track it, to avoid any confusion.

Done: #9410
